AMDGPU/GlobalISel: Uniformity info based regbankselect

petar-avramovic opened the pull request:
RFC for approach in rewriting reg-bank-selection using machine uniformity info analysis.
Implemented only for one opcode in this patch (G_FADD).
Current proposal in to pre-select register banks on dst registers using MUI before regbankselection starts.
Then using pre-assigned dst reg banks, instead of looking into register banks for inputs, select register banks for instructions so that available machine instructions could be instruction-selected.

Switch to uniformity info based regbank select could go step by step for groups of opcodes with similar operands since it does not break anything (until we get to phis).

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegBankSelect.cpp:
```diff
@@ -68,7 +71,24 @@ bool AMDGPURegBankSelect::runOnMachineFunction(MachineFunction &MF) {
+      switch (MI.getOpcode()) {
+      case AMDGPU::G_FADD: {
+        Register Dst = MI.getOperand(0).getReg();
+        if (Uniformity.isUniform(Dst)) {
+          MRI->setRegBank(Dst, RBI->getRegBank(AMDGPU::SGPRRegBankID));
+        } else {
+          MRI->setRegBank(Dst, RBI->getRegBank(AMDGPU::VGPRRegBankID));
+        }
```
`{}` isn't needed

Also if this is going to be expanded for many opcodes, might be worth putting `RBI->getRegBank(AMDGPU::SGPRRegBankID)` and the VGPR one into variables to slightly reduce verbosity

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegisterBankInfo.cpp:
```diff
@@ -695,58 +695,108 @@ static LLT getHalfSizedType(LLT Ty) {
-    auto Unmerge = B.buildUnmerge(S32, Src);
-    for (unsigned i = 0; i < NumParts; ++i)
-      SrcParts.push_back(Unmerge.getReg(i));
+  for (unsigned i = 0; i < NumElts; ++i) {
```
`i` -> `I`

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegisterBankInfo.cpp:
```diff
@@ -695,58 +695,108 @@ static LLT getHalfSizedType(LLT Ty) {
+void AMDGPURegisterBankInfo::buildReadFirstLaneB32(MachineIRBuilder &B,
+                                                   Register SgprDst,
+                                                   Register VgprSrc) const {
+  MachineRegisterInfo &MRI = *B.getMRI();
```
nit: can MRI & B be store inside `RegisterBankInfo` to reduce the need to pass them around that much?

petar-avramovic replied:
Since we mention MachineIRBuilder here how do we want to implement it?
I guess we actually want cse builder.
What about observers, do we add them in base class or is it something targets should do if they want?

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegisterBankInfo.cpp:
```diff
@@ -695,58 +695,108 @@ static LLT getHalfSizedType(LLT Ty) {
-    Register DstPart = MRI.createVirtualRegister(&AMDGPU::SReg_32RegClass);
-    MRI.setType(DstPart, NumParts == 1 ? Ty : S32);
+  B.buildUnmerge(VgprSrcParts, VgprSrc);
+  for (unsigned i = 0; i < NumElts; ++i) {
+    buildReadFirstLaneB32(B, SgprDstParts[i], VgprSrcParts[i]);
+  }
```
`i` -> `I` 
`{}` not needed

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegisterBankInfo.cpp:
```diff
@@ -695,58 +695,108 @@ static LLT getHalfSizedType(LLT Ty) {
 
-    B.buildInstr(AMDGPU::V_READFIRSTLANE_B32, {DstPart}, {SrcPart});
+  if (Ty == S32 || Ty == LLT::pointer(3, 32)) {
+    return buildReadFirstLaneB32(B, SgprDst, VgprSrc);
+  }
 
-    DstParts.push_back(DstPart);
+  if (Ty == S64 || Ty == LLT::pointer(0, 64) || Ty == LLT::pointer(1, 64)) {
+    return buildReadFirstLaneSequenceOfB32(B, SgprDst, VgprSrc, 2);
   }
```
`{}` not needed x2

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegisterBankInfo.cpp:
```diff
@@ -1015,10 +1065,23 @@ void AMDGPURegisterBankInfo::constrainOpWithReadfirstlane(
+  if (DstBank != &AMDGPU::VGPRRegBank)
+    buildReadFirstLaneDst(B, MI);
+
+  return;
```
extra `return`

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegisterBankInfo.cpp:
```diff
@@ -3355,6 +3429,28 @@ AMDGPURegisterBankInfo::getDefaultMappingVOP(const MachineInstr &MI) const {
+  OpdsMapping[0] =
+      getPreAssignedOpMapping(MI.getOperand(0).getReg(), MRI, *TRI);
+
+  for (unsigned i = 1, e = MI.getNumOperands(); i != e; ++i) {
```
`i` -> `I`, `e` -> `E`

petar-avramovic replied:
I see both lower case and upper case `i` and `I` used equally often. Is there a particular reason why to use upper case `I`?

Pierre-vh replied:
Where in the codebase do you see `i` ? A quick search for `i =` shows no result for me
In any case, coding style says variables should be uppercase so I think it's better to follow the coding style

petar-avramovic replied:
From /llvm/lib/
`unsigned I =` 1185 `int I =` 276
`unsigned i =` 2833 `int i =` 551

Pierre-vh replied:
Ah right, sure, but I still think that the coding style should be followed: https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly
So `I` is technically correct (the best kind of correct :) )

OTOH, this is really a microscopic issue (after all this is just a loop count) and I'm not going to reject the PR over that so do whatever you prefer in the end

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegisterBankInfo.cpp:
```diff
@@ -3507,6 +3603,22 @@ AMDGPURegisterBankInfo::getVGPROpMapping(Register Reg,
+  assert(Bank);
+  unsigned BankId = Bank->getID();
+  unsigned Size = getSizeInBits(Reg, MRI, TRI);
+  if (Size != 1)
+    assert(BankId == AMDGPU::SGPRRegBankID || BankId == AMDGPU::VGPRRegBankID);
+  else
+    assert(BankId == AMDGPU::SGPRRegBankID || BankId == AMDGPU::VCCRegBankID);
```
I'd prefer this, but it's really a small nit so pick whatever you like:
```
assert(BankId == AMDGPU::SGPRRegBankID || BankId == (Size == 1 ? AMDGPU::VCCRegBankID :  AMDGPU::VGPRRegBankID));
```

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegisterBankInfo.cpp:
```diff
@@ -3623,6 +3735,23 @@ AMDGPURegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
+  // Switch for uniformity info based regbank selection.
+  // Does not inspect register bank on incoming operands.
+  switch (MI.getOpcode()) {
+  case AMDGPU::G_FADD: {
```
Does this switch need to match what's in `AMDGPURegBankSelect` & `applyMappingImpl` ? If so, please add a comment to all the switches to note that they must be synced.

Pierre-vh reviewed (changes requested):
extra small nit: Title should use `[AMDGPU][GlobalISel]` just to make git log easier to search

Also description is a bit confusing, do you intend to add all the opcodes in this commit if the direction is good to go?
Or will this only implement G_FADD and others will come in other patches?
In any case please update the description to reflect the intention

nhaehnle commented:
Is somebody else able to take on this review please?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankSelect.cpp:
```diff
@@ -68,7 +71,26 @@ bool AMDGPURegBankSelect::runOnMachineFunction(MachineFunction &MF) {
+
+  // Switch for uniformity info based regbank selection. Pre-selects register
+  // bank on dst registers using machine uniformity analysis.
+  // Keep in sinc with switches in getInstrMapping and applyMappingImpl.
```
```suggestion
  // Keep in sync with switches in getInstrMapping and applyMappingImpl.
```

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegisterBankInfo.cpp:
```diff
@@ -700,58 +700,105 @@ static LLT getHalfSizedType(LLT Ty) {
+    Register SgprDst32 = MRI.createGenericVirtualRegister(S32);
+    MRI.setRegBank(SgprDst32, AMDGPU::SGPRRegBank);
+
+    B.buildAnyExt(VgprSrc32, VgprSrc);
```
Fold the register creation into the build calls, can set the bank after

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegisterBankInfo.cpp:
```diff
@@ -700,58 +700,105 @@ static LLT getHalfSizedType(LLT Ty) {
 
-  if (Bits == 32)
-    return DstParts[0];
+  if (Ty == S32 || Ty == LLT::pointer(3, 32))
```
Don't use hardcoded pointer type. This isn't the only 32-bit pointer type either. Plus missing <2 x s16>

petar-avramovic committed:
AMDGPU/GlobalISel: refactor build readfirstlane helpers

Refactor helpers that build readfirstlane for input registers.
Required by upcoming patches thet need to build readfirstlane
for output registers.

petar-avramovic committed:
AMDGPU/GlobalISel: uniformity analysis based register bank selection

Current algorithm only considers register banks for inputs but does
take control flow into account.
This is wrong in cases where inputs are uniform (in sgpr) but because
of divergent control flow instruction is divergent and should use vgpr
instead of sgpr register banks. Most notable example are phis.
Also in cases where only available machine instruction uses vgpr
registers uniform instructions end up using vgpr register banks.
Start with simple implementation for G_FADD.
Pre-select register bank for destination register using machine
uniformity analysis info. Then select register banks that would allow
selection of available machine instructions.
For G_FADD vgpr machine instruction is available on all targets but
sgpr version is not. When there is no sgpr version assign vgpr register
banks and move vgpr destination to sgpr using readfirstlane.
